WORKSPACE

```diff
@@ -192,25 +192,48 @@ http_archive(
 
 http_archive(
     name = "rules_python",
-    sha256 = "aa96a691d3a8177f3215b14b0edc9641787abaaa30363a080165d06ab65e1161",
-    url = "https://github.com/bazelbuild/rules_python/releases/download/0.0.1/rules_python-0.0.1.tar.gz",
+    # TODO Fix bazel linter to support hashes for release tarballs.
+    #
+    # sha256 = "94750828b18044533e98a129003b6a68001204038dc4749f40b195b24c38f49f",
+    strip_prefix = "rules_python-0.21.0",
+    url = "https://github.com/bazelbuild/rules_python/releases/download/0.21.0/rules_python-0.21.0.tar.gz",
 )
 
+load("@rules_python//python:repositories.bzl", "py_repositories")
+
+py_repositories()
+
+load("@rules_python//python:repositories.bzl", "python_register_toolchains")
+
+python_register_toolchains(
+    name = "python3_8",
+    python_version = "3.8",
+)
+
+load("@python3_8//:defs.bzl", "interpreter")
+load("@rules_python//python:pip.bzl", "pip_parse")
+
+pip_parse(
+    name = "pip_deps",
+    python_interpreter_target = interpreter,
+    requirements_lock = "//:tools/build/bazel/requirements.txt",
```

Comment:
Just curious if Bazel has any caching mechanism when installing these dependencies?

Reply:
Good question! And yes, it does! So of course there's local caching: if you fetch it once, it will be available in the local on disk cache and won't need to be fetched again unless it changes.

`bazel clean` has an `--expunge` flag that wipes dependencies like a Python runtime. I believe that `bazel clean` on its own would not evict that.

But your question is more than that. I think you're also asking if there are ways of caching a dependency like this so that it won't have to be fetched by every CI worker. The answer here is also yes. Let me add some flavor though.

Note that I believe that the weight of this dependency is smaller than the weight of MKL. At the minimum, it takes me much longer to fetch MKL in Bazel than the Python interpreter.

Another thing I'd like to point out is that these Python interpreters are ultimately sourced from github as well! So we remain with the same shared fate with everything else github in CI. No new risk, really.

Anyhow, I believe you can fetch this eagerly with `bazel fetch`. So we could theoretically pre-seed our images with that.

Another thing to consider is that Bazel also has remote caching options. So I believe one thing we could do would be to have, as part of our viable/strict validation or trunk jobs a Bazel run that writes to a remote cache that we manage on AWS, and then other CI jobs can read from it. I *think* that this would apply to external dependencies, but I'd have to do some research to verify this.

Such a remote cache could actually speed up the Bazel jobs in CI altogether. And doesn't need to be restricted to viable/strict jobs.